fw/tls.c

```diff
@@ -957,15 +957,16 @@ tfw_tls_over(TlsCtx *tls, int state)
 {
 	int sk_proto = ((SsProto *)tls->sk->sk_user_data)->type;
 	TfwConn *conn = (TfwConn*)tls->sk->sk_user_data;
+	int r;
 
 	if (state == TTLS_HS_CB_FINISHED_NEW
 	    || state == TTLS_HS_CB_FINISHED_RESUMED)
 		TFW_INC_STAT_BH(serv.tls_hs_successful);
 
 	if (TFW_FSM_TYPE(sk_proto) == TFW_FSM_H2 &&
-	    tfw_h2_context_init(tfw_h2_context_unsafe(conn))) {
+	    ((r = tfw_h2_context_init(tfw_h2_context_unsafe(conn))))) {
 		    T_ERR("cannot establish a new h2 connection\n");
-		    return T_DROP;
+		    return r;
```

Comment:
Just wandering why we can't do simply `return T_BAD`, it anyway will be "casted" to T_BAD in `tfw_tls_connection_recv()`?

Reply:
It seems that it's better to return DROP_WIRH_FIN for all types of internal tls errors if (tls is noit finished). There is no sense to send tcp_shutdown if there is no pending error responses.